app/services/partnerships/report.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Partnerships
+  class Report
+    CHALLENGE_WINDOW = 14.days.freeze
+    REMINDER_EMAIL_DELAY = 7.days.freeze
+
+    def self.call(**args)
+      new(**args).call
+    end
+
+    def initialize(cohort_id:, school_id:, lead_provider_id:, delivery_partner_id:)
+      @cohort_id = cohort_id
+      @school_id = school_id
+      @lead_provider_id = lead_provider_id
+      @delivery_partner_id = delivery_partner_id
+    end
+
+    def call
+      Partnership.transaction do
+        partnership = Partnership.find_or_initialize_by(
+          school_id: school_id,
+          cohort_id: cohort_id,
+          lead_provider_id: lead_provider_id,
+        )
+
+        partnership.challenge_reason = partnership.challenged_at = nil
+        partnership.delivery_partner_id = delivery_partner_id
+        partnership.pending = school_cohort.core_induction_programme?
+        partnership.challenge_deadline = CHALLENGE_WINDOW.from_now
+        partnership.report_id = SecureRandom.uuid
+        partnership.save!
+
+        partnership.event_logs.create!(
+          event: :reported,
+        )
+
+        PartnershipNotificationService.new.delay.notify(partnership)
+        PartnershipReminderJob.set(wait: REMINDER_EMAIL_DELAY).perform_later(
+          partnership: partnership,
+          report_id: partnership.report_id,
+        )
+
+        if partnership.pending?
+          PartnershipActivationJob.new.delay(run_at: partnership.challenge_deadline).perform(
+            partnership: partnership,
+            report_id: partnership.report_id,
+          )
+        end
```

Comment:
Instead of using `report_id` to make the jobs not do their job when report id changes, can we wipe the future jobs for the partnership when we re-report the partnership?

Reply:
That would be ideal, but is not that straight forward - delayed job is not made for querying. I am happy to hack around with that in test (remember my DJ matchers?), but I would not touch this in production code if the simpler solution is available. Delayed job `delay` does yield the DJ object so we could in theory store the reference, but scheduling AJ job does not. The problematic thing would be to hold a list of job ids to delete associated with partnership record - this would potentially require either a new table (database driven active job!) or a new column with an array of job ids to remove on re-reporting. Will need to think about this a bit more, maybe there is (pretty much for sure) a solution I'm missing

Reply:
Oh yeah, I did forget that the partnership would only be retrievable from some horribly serialized delayed job. Ugh. Alright, I think I prefer having `report_id` over that.